### PR TITLE
pam_lastlog: fix file locking

### PR DESCRIPTION
**Pull request by scholarsmate**
Fixed 2 instances where file locks were not being enforced.

### CODE REVIEW

File: modules/pam_lastlog/pam_lastlog.c [hunk 1]
```diff
@@ -278,10 +280,16 @@ last_login_read(pam_handle_t *pamh, int announce, int last_fd, uid_t uid, time_t
...
+            D(("locking %s failed", _PATH_LASTLOG));
+            pam_syslog(pamh, LOG_ERR,
+                       "file %s is locked/read", _PATH_LASTLOG);
+            return PAM_SERVICE_ERR;
+        }
         D(("locking %s failed..(waiting a little)", _PATH_LASTLOG));
-	pam_syslog(pamh, LOG_WARNING,
-		   "file %s is locked/read", _PATH_LASTLOG);
+        pam_syslog(pamh, LOG_WARNING,
+                   "file %s is locked/read, retrying", _PATH_LASTLOG);
```
Comment (ldv-alt) [inline modules/pam_lastlog/pam_lastlog.c:292]: I'd say that in the new context the message about retrying is not a warning anymore,
so LOG_INFO would be more suitable.

File: modules/pam_lastlog/pam_lastlog.c [hunk 2]
```diff
@@ -426,9 +435,15 @@ last_login_write(pam_handle_t *pamh, int announce, int last_fd,
...
+    while (fcntl(last_fd, F_SETLK, &last_lock) < 0) {
+        if (0 == --lock_retries) {
+            D(("locking %s failed", _PATH_LASTLOG));
+            pam_syslog(pamh, LOG_ERR,
+                       "file %s is locked/write", _PATH_LASTLOG);
+            return PAM_SERVICE_ERR;
+        }
 	D(("locking %s failed..(waiting a little)", _PATH_LASTLOG));
-	pam_syslog(pamh, LOG_WARNING, "file %s is locked/write", _PATH_LASTLOG);
+	pam_syslog(pamh, LOG_WARNING, "file %s is locked/write, retrying", _PATH_LASTLOG);
```
Comment (ldv-alt) [inline modules/pam_lastlog/pam_lastlog.c:446]: Likewise.

**Review by ldv-alt — CHANGES_REQUESTED** [review]

**Comment by t8m** [comment]
I am not sure this is a right change. This can cause a DoS when logging in. The lastlog reading is not something that must absolutely be done under a lock.

**Comment by scholarsmate** [comment]
> I am not sure this is a right change. This can cause a DoS when logging in. The lastlog reading is not something that must absolutely be done under a lock.

There are 2 lock requests here, one for reading, and another for writing.  There are many opportunities in PAM to create DoS attacks, but I'm not seeing too much risk here with these file locks.  I'm trying to avoid a very real chance of file corruption or incorrect read.  If you're worried about returning PAM_SERVICE_ERR if we fail to get a file lock after 3 retries, I believe that can be best mitigated through the PAM configuration for lastlog.

**Comment by scholarsmate** [comment]
I made the changes requested several days ago.  Is there anything else?

**Comment by ldv-alt** [comment]
Note that this PR fails to pass the tests.

**Comment by scholarsmate** [comment]
Noted.  It was a whitespace check, which is now fixed.

**Comment by scholarsmate** [comment]
Is there anything else standing in the way for this PR?

**Comment by ldv-alt** [comment]
> Is there anything else standing in the way for this PR?

Since Tomáš @t8m raised objections, they have to be either addressed or withdrawn.

**Comment by t8m** [comment]
I would be OK with the write lock failure being fatal. Not so much with the read side.

**Comment by scholarsmate** [comment]
> I would be OK with the write lock failure being fatal. Not so much with the read side.

@t8m, pushed a change so that the read side is now non-fatal to hopefully address your concern about the possible DoS with the file locks.

File: modules/pam_lastlog/pam_lastlog.c [hunk 3]
```diff
@@ -278,11 +278,18 @@ last_login_read(pam_handle_t *pamh, int announce, int last_fd, uid_t uid, time_t
...
 
-    if (fcntl(last_fd, F_SETLK, &last_lock) < 0) {
+    while (fcntl(last_fd, F_SETLK, &last_lock) < 0) {
+        if (0 == --lock_retries) {
+            /* read lock failed, proceed anyway to avoid possible DoS */
+            D(("locking %s failed", _PATH_LASTLOG));
+            pam_syslog(pamh, LOG_INFO,
+                       "file %s is locked/read, proceeding anyway",
+                       _PATH_LASTLOG);
+        }
```
Comment (t8m) [inline modules/pam_lastlog/pam_lastlog.c:288]: You're missing a `break;` here to break out of the loop.

Comment (scholarsmate) [reply modules/pam_lastlog/pam_lastlog.c:289]: You're right, `break;` statement added.

**Review by t8m — APPROVED** [review]

**Comment by scholarsmate** [comment]
@ldv-alt, is there anything else needed on this PR?

**Comment by ldv-alt** [comment]
> @ldv-alt, is there anything else needed on this PR?

Please make sure the indentation is consistent.

Also, someone would have to rebase this on top of the master branch, it should be a single commit with a descriptive commit message.

**Comment by scholarsmate** [comment]
> Please make sure the indentation is consistent.

The indentation of this file was inconsistent before this change, so it's not really easy to figure out what is proper and what isn't.  I'd recommend running a formatter (e.g., clang-tidy) to make it all consistent as a separate pull request, which I'd be happy to provide.

> Also, someone would have to rebase this on top of the master branch, it should be a single commit with a descriptive commit message.

Let me know what you want for correcting the indentation, and I'll make the changes.  I'll then rebase the branch on master, squash the commits, write up a good description, then do a force push on this branch so that it becomes a single commit for testing, re-approval, and merge.

**Comment by scholarsmate** [comment]
@ldv-alt, the commits have been rebased and squashed, and the commit message now has a better description of the changes.  All the tests are passing in my fork.  Hope this helps to get it merged.

**Review by ldv-alt — APPROVED** [review]

**Pull request merged**